src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -2668,5 +2668,36 @@ internal bool IsTypeMissing(WellKnownType type)
         }
         
         internal abstract bool IsIOperationFeatureEnabled();
+
+        /// <summary>
+        /// Given a <see cref="Diagnostic"/> reporting unreferenced <see cref="AssemblyIdentity"/>s, returns
+        /// the actual <see cref="AssemblyIdentity"/> instances that were not referenced.
+        /// </summary>
+        public ImmutableArray<AssemblyIdentity> GetUnreferencedAssemblyIdentities(Diagnostic diagnostic)
+        {
+            if (diagnostic == null)
+            {
+                throw new ArgumentNullException(nameof(diagnostic));
+            }
+
+            if (!UnreferencedAssemblyIdentityDiagnosticCodes.Contains(diagnostic.Code))
+            {
+                return ImmutableArray<AssemblyIdentity>.Empty;
+            }
+
+            var builder = ArrayBuilder<AssemblyIdentity>.GetInstance();
+
+            foreach (var argument in diagnostic.Arguments)
+            {
+                if (argument is AssemblyIdentity id)
+                {
+                    builder.Add(id);
+                }
+            }
+
+            return builder.ToImmutableAndFree();
+        }
+
+        internal abstract ImmutableArray<int> UnreferencedAssemblyIdentityDiagnosticCodes { get; }
```

Comment:
>internal abstract ImmutableArray<int> UnreferencedAssemblyIdentityDiagnosticCodes { get; } [](start = 8, length = 90)

It feels like it would be better to have an API that returns true/false for a specific error code.